Introduce cuda_p2p based fused_all_gather_matmul and fused_matmul_reduce_scatter

yifuwang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #127455
* #127454
* #126599
* #126598
* __->__ #126634



cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

Commit:
Introduce cuda_p2p based fused_all_gather_matmul and fused_matmul_reduce_scatter

[ghstack-poisoned]

Commit:
Update on "Introduce cuda_p2p based fused_all_gather_matmul and fused_matmul_reduce_scatter"


## Context
See context [here](https://github.com/pytorch/pytorch/pull/122163).

## This PR
Introduces `cuda_p2p` based `fused_all_gather_matmul` and `fused_matmul_reduce_scatter` dispatcher ops which performs micro-pipelining TP for `all-gather -> matmul` and `matmul -> reduce-scatter` respectively.

Fusion vs. decomposition - in principle, the micro-pipelining is achieved via decomposition. However, in practice, today Inductor can't deal with the decomposed patterns well. So instead performing decomposition in Inductor, we fuse the patterns to be decomposed and dispatch them to corresponding operators that handle decomposition + micropipelining.

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

Chillee commented on torch/distributed/_cuda_p2p/__init__.py:
```diff
@@ -121,3 +121,360 @@ def get_p2p_buffer_size(group: c10d.ProcessGroup) -> int:
+
+    with torch.cuda.stream(backend.stream()):
+        local_p2p_buf.copy_(shard)
+        work = backend.intra_node_barrier()
```
Why is this the only one that needs a barrier?

yifuwang replied:
After this barrier, the condition "shard `i` is available in rank `i`'s p2p buffer" will be met, after which each rank can "pull" from remote ranks without additional synchronization. We have another barrier at the end, which ensures that "all ranks have finished reading from remote ranks" so that the p2p buffers can be safely reused.

Chillee commented on torch/distributed/_cuda_p2p/__init__.py:
```diff
@@ -121,3 +121,360 @@ def get_p2p_buffer_size(group: c10d.ProcessGroup) -> int:
+    group_size = group.size()
+    rank = group.rank()
+
+    out_chunks = output.chunk(group_size)
```
Does this approach in higher memory usage? Since we need to materialize both the input and output at the same time?

yifuwang replied:
It is higher compared to the online accumulation approach where each rank only needs to maintain the `reduce_scatter` result. It is not higher compare to the baseline which materializes the matmul output anyway.

We do require a p2p buffer 2x the size of the `reduce_scatter` result to allow for more parallelism.

Chillee reviewed (approved):
I do feel like we should be able to provide a higher level API for this 🤔 It would be nice if it could be the same API for both allgather and reduce_scatter.

wanchaol reviewed (approved):
looks awesome!

Commit:
Update on "Introduce cuda_p2p based fused_all_gather_matmul and fused_matmul_reduce_scatter"


## Context
See context [here](https://github.com/pytorch/pytorch/pull/122163).

## This PR
Introduces `cuda_p2p` based `fused_all_gather_matmul` and `fused_matmul_reduce_scatter` dispatcher ops which performs micro-pipelining TP for `all-gather -> matmul` and `matmul -> reduce-scatter` respectively.

Fusion vs. decomposition - in principle, the micro-pipelining is achieved via decomposition. However, in practice, today Inductor can't deal with the decomposed patterns well. So instead performing decomposition in Inductor, we fuse the patterns to be decomposed and dispatch them to corresponding operators that handle decomposition + micropipelining.

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

Commit:
Update on "Introduce cuda_p2p based fused_all_gather_matmul and fused_matmul_reduce_scatter"


## Context
See context [here](https://github.com/pytorch/pytorch/pull/122163).

## This PR
Introduces `cuda_p2p` based `fused_all_gather_matmul` and `fused_matmul_reduce_scatter` dispatcher ops which performs micro-pipelining TP for `all-gather -> matmul` and `matmul -> reduce-scatter` respectively.

Fusion vs. decomposition - in principle, the micro-pipelining is achieved via decomposition. However, in practice, today Inductor can't deal with the decomposed patterns well. So instead performing decomposition in Inductor, we fuse the patterns to be decomposed and dispatch them to corresponding operators that handle decomposition + micropipelining.

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

yifuwang commented on torch/distributed/_cuda_p2p/__init__.py:
```diff
@@ -121,3 +121,360 @@ def get_p2p_buffer_size(group: c10d.ProcessGroup) -> int:
+    ag_shape = list(A_shard.shape)
+    ag_shape[gather_dim] *= group_size
+    ag_out = A_shard.new_empty(ag_shape)
+    return ag_out, [ag_out @ B for B in Bs]
```
Yeah we are calling the meta kernels to deduce device, shape, and strides.

Commit:
Update on "Introduce cuda_p2p based fused_all_gather_matmul and fused_matmul_reduce_scatter"


## Context
See context [here](https://github.com/pytorch/pytorch/pull/122163).

## This PR
Introduces `cuda_p2p` based `fused_all_gather_matmul` and `fused_matmul_reduce_scatter` dispatcher ops which performs micro-pipelining TP for `all-gather -> matmul` and `matmul -> reduce-scatter` respectively.

Fusion vs. decomposition - in principle, the micro-pipelining is achieved via decomposition. However, in practice, today Inductor can't deal with the decomposed patterns well. So instead performing decomposition in Inductor, we fuse the patterns to be decomposed and dispatch them to corresponding operators that handle decomposition + micropipelining.

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

Commit:
Update on "Introduce cuda_p2p based fused_all_gather_matmul and fused_matmul_reduce_scatter"


## Context
See context [here](https://github.com/pytorch/pytorch/pull/122163).

## This PR
Introduces `cuda_p2p` based `fused_all_gather_matmul` and `fused_matmul_reduce_scatter` dispatcher ops which performs micro-pipelining TP for `all-gather -> matmul` and `matmul -> reduce-scatter` respectively.

Fusion vs. decomposition - in principle, the micro-pipelining is achieved via decomposition. However, in practice, today Inductor can't deal with the decomposed patterns well. So instead performing decomposition in Inductor, we fuse the patterns to be decomposed and dispatch them to corresponding operators that handle decomposition + micropipelining.

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

Commit:
Update on "Introduce cuda_p2p based fused_all_gather_matmul and fused_matmul_reduce_scatter"


## Context
See context [here](https://github.com/pytorch/pytorch/pull/122163).

## This PR
Introduces `cuda_p2p` based `fused_all_gather_matmul` and `fused_matmul_reduce_scatter` dispatcher ops which performs micro-pipelining TP for `all-gather -> matmul` and `matmul -> reduce-scatter` respectively.

Fusion vs. decomposition - in principle, the micro-pipelining is achieved via decomposition. However, in practice, today Inductor can't deal with the decomposed patterns well. So instead performing decomposition in Inductor, we fuse the patterns to be decomposed and dispatch them to corresponding operators that handle decomposition + micropipelining.

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

Commit:
Update on "Introduce cuda_p2p based fused_all_gather_matmul and fused_matmul_reduce_scatter"




[ghstack-poisoned]

Commit:
Update on "Introduce cuda_p2p based fused_all_gather_matmul and fused_matmul_reduce_scatter"




[ghstack-poisoned]

yifuwang commented:
@pytorchbot merge
